// @heroku/salesforce-sdk/dist/functions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Represents invoking user.
 */
class UserContext {
    constructor(orgDomainUrl, orgId, salesforceBaseUrl, username, userId, onBehalfOfUserId) {
        this.orgDomainUrl = orgDomainUrl;
        this.orgId = orgId;
        this.salesforceBaseUrl = salesforceBaseUrl;
        this.username = username;
        this.userId = userId;
        this.onBehalfOfUserId = onBehalfOfUserId;
    }
}
exports.UserContext = UserContext;
/**
 * Respresents the context of the function invocation including objects needed by functions to
 * perform their action(s), eg ForceAPI to interact with the invoking org.
 */
class Context {
    constructor(apiVersion, userContext, logger, payloadVersion, forceApi, unitOfWork) {
        this.apiVersion = apiVersion;
        this.userContext = userContext;
        this.logger = logger;
        this.payloadVersion = payloadVersion;
        this.forceApi = forceApi;
        this.unitOfWork = unitOfWork;
    }
}
exports.Context = Context;
