// @salesforce/kit/lib/nodash/external.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaults = defaults;
exports.findKey = findKey;
exports.includes = includes;
exports.keyBy = keyBy;
exports.mapKeys = mapKeys;
exports.minBy = minBy;
exports.maxBy = maxBy;
exports.merge = merge;
exports.omit = omit;
exports.once = once;
exports.set = set;
exports.sortBy = sortBy;
exports.toNumber = toNumber;
/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
// @ts-ignore ignore the demand for typings for the locally built lodash
const _ = __importStar(require("../../vendor/lodash"));
// underlying function
function defaults(obj, ...otherArgs) {
    return _.defaults(obj, ...otherArgs);
}
/**
 * This method is like `find` except that it returns the key of the first element predicate returns truthy for
 * instead of the element itself.
 *
 * @param obj The object to search.
 * @param predicate The function invoked per iteration.
 */
function findKey(obj, predicate) {
    return _.findKey(obj, predicate);
}
/**
 * Checks if target is in collection using SameValueZero for equality comparisons. If fromIndex is negative,
 * it’s used as the offset from the end of collection.
 *
 * @param collection The collection to search.
 * @param target The value to search for.
 * @param fromIndex The index to search from.
 */
function includes(collection, target, fromIndex) {
    return _.includes(collection, target, fromIndex);
}
// underlying function
function keyBy(collection, iteratee) {
    return _.keyBy(collection, iteratee);
}
// underlying function
function mapKeys(obj, iteratee) {
    return _.mapKeys(obj, iteratee);
}
/**
 * This method is like `_.min` except that it accepts `iteratee` which is
 * invoked for each element in `array` to generate the criterion by which
 * the value is ranked. The iteratee is invoked with one argument: (value).
 *
 * @param array The array to iterate over.
 * @param iteratee The iteratee invoked per element.
 */
function minBy(collection, iteratee) {
    return _.minBy(collection, iteratee);
}
/**
 * This method is like `_.max` except that it accepts `iteratee` which is
 * invoked for each element in `array` to generate the criterion by which
 * the value is ranked. The iteratee is invoked with one argument: (value).
 *
 * @param array The array to iterate over.
 * @param iteratee The iteratee invoked per element.
 */
function maxBy(collection, iteratee) {
    return _.maxBy(collection, iteratee);
}
// underlying function
function merge(obj, ...otherArgs) {
    return _.merge(obj, ...otherArgs);
}
// underlying function
function omit(obj, ...paths) {
    return _.omit(obj, ...paths);
}
/**
 * Creates a function that is restricted to invoking `func` once. Repeat calls to the function return the value
 * of the first call. The `func` is invoked with the this binding and arguments of the created function.
 *
 * @param func The function to restrict.
 */
function once(func) {
    return _.once(func);
}
// underlying function
function set(obj, path, value) {
    return _.set(obj, path, value);
}
// underlying function
function sortBy(collection, ...iteratees) {
    return _.sortBy(collection, ...iteratees);
}
/**
 * Converts `value` to a number.
 *
 * @param value The value to process.
 *
 * ```
 * _.toNumber(3);
 * // => 3
 *
 * _.toNumber(Number.MIN_VALUE);
 * // => 5e-324
 *
 * _.toNumber(Infinity);
 * // => Infinity
 *
 * _.toNumber('3');
 * // => 3
 * ```
 */
function toNumber(value) {
    return _.toNumber(value);
}
