// @heroku/applink/dist/sdk/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerImpl = void 0;
const base_logger_1 = __importDefault(require("../utils/base-logger"));
class LoggerImpl {
    constructor(id) {
        this.properties = {
            requestId: id,
        };
    }
    error(message) {
        underlyingLogger.error(Object.assign(this.properties, { message }));
    }
    warn(message) {
        underlyingLogger.warn(Object.assign(this.properties, { message }));
    }
    info(message) {
        underlyingLogger.info(Object.assign(this.properties, { message }));
    }
    debug(message) {
        underlyingLogger.debug(Object.assign(this.properties, { message }));
    }
    trace(message) {
        underlyingLogger.trace(Object.assign(this.properties, { message }));
    }
}
exports.LoggerImpl = LoggerImpl;
const underlyingLogger = base_logger_1.default;
