// @heroku/applink/dist/sdk/unit-of-work.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitOfWorkImpl = exports.ReferenceIdImpl = void 0;
const sub_request_1 = require("./sub-request");
class ReferenceIdImpl {
    constructor(id) {
        this.id = id;
    }
    toString() {
        return this.id;
    }
    toApiString() {
        return `@{${this.id}.id}`;
    }
}
exports.ReferenceIdImpl = ReferenceIdImpl;
class UnitOfWorkImpl {
    constructor(apiVersion) {
        this._subrequests = [];
        this.referenceIdCounter = 0;
        this.apiVersion = apiVersion;
    }
    registerCreate(record) {
        const referenceId = this.generateReferenceId();
        this._subrequests.push([referenceId, new sub_request_1.CreateRecordSubRequest(record)]);
        return referenceId;
    }
    registerDelete(type, id) {
        const referenceId = this.generateReferenceId();
        this._subrequests.push([referenceId, new sub_request_1.DeleteRecordSubRequest(type, id)]);
        return referenceId;
    }
    registerUpdate(record) {
        const referenceId = this.generateReferenceId();
        this._subrequests.push([referenceId, new sub_request_1.UpdateRecordSubRequest(record)]);
        return referenceId;
    }
    get subrequests() {
        return this._subrequests;
    }
    generateReferenceId() {
        const referenceId = new ReferenceIdImpl("referenceId" + this.referenceIdCounter);
        this.referenceIdCounter = this.referenceIdCounter + 1;
        return referenceId;
    }
}
exports.UnitOfWorkImpl = UnitOfWorkImpl;
