// jsforce/module/api/apex.js
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
/**
 * @file Manages Salesforce Apex REST endpoint calls
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { registerModule } from '../jsforce';
/**
 * API class for Apex REST endpoint call
 */
export class Apex {
  /**
   *
   */
  constructor(conn) {
    _defineProperty(this, "_conn", void 0);
    _defineProperty(this, "del", this.delete);
    this._conn = conn;
  }

  /* @private */
  _baseUrl() {
    return `${this._conn.instanceUrl}/services/apexrest`;
  }

  /**
   * @private
   */
  _createRequestParams(method, path, body, options = {}) {
    const headers = typeof options.headers === 'object' ? options.headers : {};
    if (!/^(GET|DELETE)$/i.test(method)) {
      headers['content-type'] = 'application/json';
    }
    const params = {
      method,
      url: this._baseUrl() + path,
      headers
    };
    if (body) {
      params.body = _JSON$stringify(body);
    }
    return params;
  }

  /**
   * Call Apex REST service in GET request
   */
  get(path, options) {
    return this._conn.request(this._createRequestParams('GET', path, undefined, options));
  }

  /**
   * Call Apex REST service in POST request
   */
  post(path, body, options) {
    const params = this._createRequestParams('POST', path, body, options);
    return this._conn.request(params);
  }

  /**
   * Call Apex REST service in PUT request
   */
  put(path, body, options) {
    const params = this._createRequestParams('PUT', path, body, options);
    return this._conn.request(params);
  }

  /**
   * Call Apex REST service in PATCH request
   */
  patch(path, body, options) {
    const params = this._createRequestParams('PATCH', path, body, options);
    return this._conn.request(params);
  }

  /**
   * Call Apex REST service in DELETE request
   */
  delete(path, options) {
    return this._conn.request(this._createRequestParams('DELETE', path, undefined, options));
  }

  /**
   * Synonym of Apex#delete()
   */
}

/*--------------------------------------------*/
/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */
registerModule('apex', conn => new Apex(conn));
export default Apex;
